### ISSUE TITLE
"Missing self argument" error after dynamically replacing method

### ISSUE THREAD

**Issue by shiw-yang** [astral-sh/ty#350]
### Summary

In the provided example, I dynamically replace the A.f1 method with f1_replace through the f2 function. After this replacement,  ty reports an error:
No argument provided for required parameter 'self' of function 'f1' when calling a.f1(), though the code actually runs correctly.

### Steps to Reproduce:

```python

class A:
    def f1(self):
        print("f1")

    def f1_replace(self):
        print("replace")

    def f2(self):
        self.f1 = self.f1_replace


a = A() 
a.f1()  # No argument provided for required parameter `self` of function `f1`
a.f2()
a.f1()  # No argument provided for required parameter `self` of function `f1`

```

then use `ty check` show: 

```bash
$ ty check main.py
WARN ty is pre-release software and not ready for production use. Expect to encounter bugs, missing features, and fatal errors.
error[missing-argument]: No argument provided for required parameter `self` of function `f1`
  --> main.py:39:1
   |
38 | a = A()
39 | a.f1()
   | ^^^^^^
40 | a.f2()
41 | a.f1()
   |
info: Union variant `def f1(self) -> Unknown` is incompatible with this call site
info: Attempted to call union type `(bound method A.f1() -> Unknown) | (def f1(self) -> Unknown)`
info: `missing-argument` is enabled by default

error[missing-argument]: No argument provided for required parameter `self` of function `f1`
  --> main.py:41:1
   |
39 | a.f1()
40 | a.f2()
41 | a.f1()
   | ^^^^^^
   |
info: Union variant `def f1(self) -> Unknown` is incompatible with this call site
info: Attempted to call union type `(bound method A.f1() -> Unknown) | (def f1(self) -> Unknown)`
info: `missing-argument` is enabled by default

Found 2 diagnostics
```

### Version

ty 0.0.0-alpha.8

**Comment by sharkdp** [comment]
Thank you for reporting this. This is a bug in ty.

The problem here seems to be that we recognize the instance attribute assignment to `self.f1`, but since we don't understand the type of `self` (#159), we infer `Unknown` for the implicit instance attribute `f1`. We union that with the type of the `def f1` method definition, but fail to bind the `self` argument (because we think it's an instance attribute and don't invoke the descriptor protocol?).

**Comment by vprimault** [comment]
I've hit this bug as well, is there a workaround? I tried to use `cast` to force the type but only got an additional `warning[redundant-cast]` and still the same error.

**Comment by sharkdp** [comment]
> I've hit this bug as well, is there a workaround?

A rather silly workaround is to redeclare `f1` on the class body as being of type `Any`, `Callable`, or `Callable[..., None]`:
```py
class A:
    def f1(self):
        print("f1")

    f1: Any
```

**Comment by sharkdp** [comment]
Another instance of this appears with the library [pulumi](https://github.com/pulumi/pulumi/tree/master/sdk/python), see [this comment](https://github.com/astral-sh/ty/issues/2878#issuecomment-4116606373). A variant of the OP example that is more similar to the pulumi case would be:

```py
class C:
    def apply(self, name: str) -> int:
        return 0

    def some_other_method(self):
        def error(): ...
        self.apply = error()

C().apply("test")
```

which leads to two very cryptic error messages:

```
No argument provided for required parameter `name` of function `apply` (missing-argument) [Ln 10, Col 1]
Argument to function `apply` is incorrect: Expected `C`, found `Literal["test"]` (invalid-argument-type) [Ln 10, Col 11]
```


Here's the problematic line in pulumi: https://github.com/pulumi/pulumi/blob/b605630ec3db33517cd4297d41e2f69ca96c5cb8/sdk/python/lib/pulumi/output.py#L198

**Comment by sharkdp** [comment]
Maybe this issue will be solved by the planned changes in #136?

**Comment by carljm** [comment]
> Maybe this issue will be solved by the planned changes in [#136](https://github.com/astral-sh/ty/issues/136)?

I don't think so? The intention of #136 is that it will only impact annotated assignment statements. I don't see any annotated assignment statement in the OP example here (or in the pulumi example).

In particular, after #136 we will still narrow the type on the non-annotated assignment here:

```py
x: int | None = None  # here we currently infer `None` but after #136 we'll infer `int | None`
x = 1  # here we currently infer `Literal[1]` for `x`, and will continue to do so after #136
``` 

This issue adds a different wrinkle, which is that the relevant treatment of the assignment is not handling the definition in local-scope control flow, but inferring a type for an attribute of a class via looking at assignments in methods. So I think that makes it even less related to the planned change in #136.

I think for attributes we already don't look at assignments in methods if we have a declared type? So one possible change to address this issue might be to consider a method definition to be a "declared type", so we wouldn't even look at the additional assignments in methods? (Other than to validate them against that declared type, which I think happens in regular checking of the method scopes, not in attribute-type inference.)

**Comment by sharkdp** [comment]
> > Maybe this issue will be solved by the planned changes in [#136](https://github.com/astral-sh/ty/issues/136)?
> 
> I don't think so? The intention of [#136](https://github.com/astral-sh/ty/issues/136) is that it will only impact annotated assignment statements. I don't see any annotated assignment statement in the OP example here (or in the pulumi example).

I should have explained this better. We treat a `def …` function definition (in a class body) as both a declaration and a binding. And so I was thinking that if we prefer the declared type (of e.g. `apply` in the latest example) over its inferred type (which is affected by the implicit instance attribute assignment), then this might be enough to only "see" the declared type of `apply` externally, that is: when accessing `apply` through an instance of `C`.

What's actually going on is more complex, though. When we access `C().apply`, we first look up `apply` on `type(C())`, i.e. on the class `C` itself. We find the `def apply` definition and see that `apply` is a non-data descriptor. Non-data descriptors on classes are lower in priority than instance level attributes, and so we need to look those up as well. Since we don't know if `some_other_method` has ever been called, we finally build the union of those two types (the instance-level type, and the non-data descriptor type), which is, in principle, an accurate description of the runtime behavior (the `C().apply` could either refer to the original function, or the replacement).

The problematic part is that the *instance*-level type of `apply` *also* refers to the function definition (because class-body declarations are also a source of instance attribute annotations). And so we see `def apply(…)` for the type of the
instance-level type (the implicit `self.apply = …` attribute assignment of `Unknown` is compatible with that type and doesn't further affect it). But instance fallbacks do not go through the descriptor protocol! And so we see the following as the type of `C().apply`:

```py
(bound method C.apply(name: str) -> int) | (def apply(self, name: str) -> int)
```

The first part here comes from the non-data descriptor at the class level (the `def apply`, after resolving `__get__`), and the second part comes from the instance-level override (where the descriptor protocol does not apply).

A year ago, I have attempted various things to resolve this (see various PRs attached to this ticket), but I would need to go look at them again to see if there was any promising route there.

Our overly precise modeling of the descriptor protocol with partial fallbacks is certainly not helping here. But reading the above again, it seems like the most problematic step is actually the fact that we consider `def apply(…)` on the class body to be evidence for the fact that there is *an instance-level attribute* of type `def apply(…)` (the function-literal type). It's not very principled, but simply filtering out all function definitions would probably solve this problem completely.

**Comment by sharkdp** [comment]
Here's the same problem with a plain (non-function) non data descriptor:
```py
class DescriptorReturnType: ...

class NonDataDescriptor:
    def __get__(self) -> DescriptorReturnType:
        return DescriptorReturnType()

class C:
    apply: NonDataDescriptor = NonDataDescriptor()

    def some_method(self, unknown):
        self.apply = unknown

reveal_type(C().apply)  # DescriptorReturnType | NonDataDescriptor
```
https://play.ty.dev/3b615dde-b982-40a3-9e10-993d6dfaae27

**Comment by carljm** [comment]
That all makes sense! I agree that we should not consider `def apply(...)` in the class body as an instance attribute type. I think this may go beyond just functions, but maybe to any declaration-and-binding in the class body? I'm not sure that we'd ever want to consider `x: SomeType = SomeType()` in the class body, in itself, as evidence of an instance-level attribute of type `SomeType`. (If `SomeType` is not a descriptor type, it doesn't matter -- if it is a descriptor type, including it leads to the problem discussed here.)

To be clear, though, I consider all of this to be separate from #136. I don't expect to make any changes to attribute inference as part of #136, only to treatment of annotated assignment statements when inferred as a binding, which I don't think would impact any examples discussed here.

**Comment by dbellavista** [comment]
May be related, but I'd like to find a way to solve this:

```python
from torchmetrics import MetricCollection

a: MetricCollection = MetricCollection({})
a["test"].compute()
```

```
$ uv run ty check test.py
error[missing-argument]: No argument provided for required parameter `self` of function `Metric.compute`
 --> test.py:4:1
  |
4 | a["test"].compute()
  | ^^^^^^^^^^^^^^^^^^^
  |
info: Union variant `def compute(self) -> Any` is incompatible with this call site
info: Attempted to call union type `(bound method Metric.compute() -> Any) | (def compute(self) -> Any)`

Found 1 diagnostic
```
